### DISCUSSION TITLE
Building on Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by jketema** [microsoft/CCF#5963] [Q&A] [answered]
Is is possible to build CCF on Ubuntu 22.04? If so, are there instructions for that somewhere?

**Comment by achamayou** [comment, marked as answer]
@jketema the short answer is no, not without changes.

The longer answer is that all CCF targets (SGX, SNP and virtual) include support for at least verifying SGX attestations (used both internally to allow mixed networks, and exposed in the JS API). That functionality is implemented by the Open Enclave SDK (https://github.com/openenclave/openenclave/), which does not support Ubuntu 22.04.

Should you want to try to do this, there are two options:

1. Try to produce a build of Open Enclave on Ubuntu 22.04
2. Remove SGX attestation support (and the oehost dependency) in CCF, perhaps behind a build flag. The resulting build ought to work on 22.04.

Option 2. is probably something we will look at sooner or later, because while SGX attestation support is necessary to support migration of existing network, we do not intend to continue support once most deployments have moved to SNP.

**Reply by jketema** [reply]
Thanks for your answer. Then it seems that I will need to set up a Ubuntu 20.04 environment for the time being.

**Reply by achamayou** [reply]
@jketema depending on what you want to do, you may find the devcontainer running in a Codespace to be good option that requires little initial setup. The limitation will be that you cannot use SGX (Codespces have no SGX-enabled SKUs), and can only build and run Virtual.

**Reply by jketema** [reply]
Thanks. Yeah, a Codespace might offer a solution. Concretely what I'm trying to do is debug a CodeQL issue. We're trying to upgrade the CodeQL C/C++ parser and we're seeing a performance regression on CCF.

**Reply by achamayou** [reply]
@jketema so there is a build flag to remove OE usage on virtual/SNP builds: `REQUIRE_OPENENCLAVE` (e.g. `cmake -GNinja -DCMAKE_BUILD_TYPE=Debug -DREQUIRE_OPENENCLAVE=OFF -DCOMPILE_TARGET=virtual ..`), and I've confirmed that works on 20.04 just now. What there isn't is a corresponding flag for the ansible playbooks to not install OE + dependencies, I will add that, and then try to rev up the CodeQL CI to 22.04. I would be surprised if there were too many other changes necessary.

**Reply by achamayou** [reply]
@jketema https://github.com/microsoft/CCF/tree/codeql_2204 has a few changes towards building on 22.04, but doesn't yet, I will need to find more time later this week (history of errors under https://github.com/microsoft/CCF/actions/runs/7698958331/job/20979472752).

One thing that definitely does not work is building with clang-11 on 22.04, the compiler just crashes during a bit of template resolution.

**Reply by jketema** [reply]
Thanks. With the help of some codespace trickery, I've been able to reduce the problem now.

**Comment by achamayou** [comment]
Sorry hit the wrong button, happy to continue the discussion of course.
